engines/docker/image_cache.go

```diff
@@ -0,0 +1,103 @@
+package dockerengine
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
+	"github.com/taskcluster/taskcluster-worker/runtime/caching"
+	"github.com/taskcluster/taskcluster-worker/runtime/gc"
+)
+
+const dockerPullImageInactivityTimeout = 30 * time.Second
+
+type imageResource struct {
+	Repository string
+	Tag        string
+	Size       uint64
+	dispose    atomics.Once
+}
+
+type cacheContext struct {
+	context.Context
+}
+
+func newCacheContext(ctx context.Context) *cacheContext {
+	if ctx != nil {
+		return &cacheContext{
+			Context: ctx,
+		}
+	}
+	return nil
+}
+
+func (ctx *cacheContext) Progress(description string, percent float64) {
+	debug("progress: %s %f", description, percent)
+}
+
+func (im *imageResource) MemorySize() (uint64, error) {
+	return im.Size, nil
+}
+
+func (im *imageResource) DiskSize() (uint64, error) {
+	return 0, gc.ErrDisposableSizeNotSupported
+}
+
+func (im *imageResource) Dispose() error {
+	im.dispose.Do(func() {
+		client, err := docker.NewClientFromEnv()
+		if err != nil {
+			debug("%v", err)
+			return
+		}
+		err = client.RemoveImage(buildImageName(im.Repository, im.Tag))
+		if err != nil {
+			debug("%v", err)
+		}
+	})
+	return nil
+}
+
+func imageConstructor(ctx caching.Context, opts interface{}) (caching.Resource, error) {
+	options := opts.(imageType)
+	client, err := docker.NewClientFromEnv()
+	if err != nil {
+		return nil, err
+	}
+
+	debug("pulling image %s %s", options.Repository, options.Tag)
+	err = client.PullImage(docker.PullImageOptions{
+		Repository:        options.Repository,
+		Tag:               options.Tag,
+		InactivityTimeout: dockerPullImageInactivityTimeout,
+		Context:           ctx,
+	}, docker.AuthConfiguration{})
+	if err != nil {
+		debug("error pulling image %v", err)
+		return nil, err
```

Comment:
I LOVE that you remembered to convert this into a `NewMalformedPayloadError` later...
I'm thinking it might be neat to do it here... that way we just return whatever error we get from pulling the image...

Also if we have an internal error of some sort we can report it and return `ErrNonFatalInternalError` right now, returning any error just becomes a malformed payload error... (which definitely should be the default).

Mostly, I just like having the option of making better errors in the future... In particularly if we ever get around to read the status feed from docker. it might include things like errors with more specific messages, like if image is private, etc. (let's not do that now :) we can have TODO for that)